// @keystonejs/fields-content/views/editor/blocks/blockquote/dist/fields-content.esm.js
import _extends from '@babel/runtime/helpers/esm/extends';
import { jsx } from '@emotion/core';
import { h as hasAncestorBlock } from '../../../../../dist/utils-d86d79bf.esm.js';
import { QuoteIcon } from '@primer/octicons-react';
import { T as ToolbarButton } from '../../../../../dist/toolbar-components-6b3a9f07.esm.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import '@arch-ui/theme';
import '@arch-ui/color-utils';
import '@arch-ui/tooltip';
import '@arch-ui/typography';

let type = 'blockquote';
function ToolbarElement({
  editor,
  editorState
}) {
  let hasBlockquote = hasAncestorBlock(editorState, type);
  return jsx(ToolbarButton, {
    isActive: hasBlockquote,
    icon: jsx(QuoteIcon, null),
    label: "Blockquote",
    onClick: () => {
      if (hasBlockquote) {
        editor.unwrapBlock(type);
      } else {
        editor.wrapBlock(type);
      }

      editor.focus();
    }
  });
}
function Node({
  attributes,
  children
}) {
  return jsx("blockquote", _extends({}, attributes, {
    css: {
      borderLeft: '4px solid #eee',
      color: '#666',
      fontStyle: 'italic',
      margin: 0,
      marginBottom: '1em',
      paddingLeft: '1em'
    }
  }), children);
}

export { Node, ToolbarElement, type };
